2_Code/src/uart.c

```diff
@@ -1,20 +1,95 @@
+/* 
+ * Date: 03/30/2024
+ * Course number: EE 2361
+ * Term: Spring
+ * Lab/assignment number: Final Project
+ * Short Program Description: UART Library to write and read to Grove - GPS Air530.
+ */
+
 #include "xc.h"
-#include "uart.h"
+#include "UART.h"
+#include <stdio.h>
+#include <stdlib.h>
+
+volatile unsigned char buffer[64]; // Circular buffer to store received data
+volatile unsigned char front = 0; // Index for adding data to front of the buffer
+volatile unsigned char back = 0; // Index for reading data from the back of the buffer
+
+void __attribute__((__interrupt__,__auto_psv__)) _U1RXInterrupt(void)
+{
+    // Updates the circular buffer with received UART data
+    // Clears the UART receive interrupt flag
+    // front: index for storing data in the buffer
+    // U1RXREG: received data from UART
+    
+    IFS0bits.U1RXIF = 0;
+    buffer[front++] = U1RXREG;
+    front &= 63;
+}
+
+void send_GPS_Str_command(unsigned char* s)
+{
+    // Sends a String to the GPS module using send_GPS_Char_command function
+    // s: the string to be transmitted
+    
+    int i = 0;
+    while (s[i] != '\0') 
+    {
+        send_GPS_Char_command(s[i]);
+        i++;
+    }
+    
+}
 
-void initUART(int baudRate)
+void send_GPS_Char_command(unsigned char command)
 {
-    // Set pins to digital in/out
+    // Sends a character to the GPS module via UART
+    // command: the character to be transmitted
+
+    while (!IFS0bits.U1TXIF);
+    IFS0bits.U1TXIF = 0;
+    U1TXREG = command;
+}
 
-    // Set up UART (not complete)
-    U1MODE = 0x0;        // 8-bit, no parity, 1 stop bit
-    U1MODEbits.BRGH = 0; // Not high speed
-    U1BRG = (unsigned int)FCY / (16 * baudRate) - 1; // Calculate baud (should be 103)
-    U1MODEbits.ON = 1;   // Enable UART
+unsigned char get_GPS_data(void)
+{
+    // Retrieves a single NMEA character from the circular buffer
+    // Waits until data is available (front is different from back)
+    // NMEA: the retrieved character
+    
+    unsigned char NMEA;
 
-    // Peripheral Pin Select
+    while (front == back) ;
+    NMEA = buffer[back++];
+    back &= 63;
+    return NMEA;
 }
 
-void sendPackage(char data[])
+void init_UART(baudRate)
 {
-    // Write to TX data stream using other 'sendChar' function
+    // Initializes UART communication settings
+    // Configures UART module, pins, and interrupt flags
+
+    CLKDIVbits.RCDIV = 0;
+    AD1PCFG = 0x9fff;  // For digital I/O.
```

Comment:
Although it is more work, we should clarify which pins we are using for UART as opposed to overwriting all configurations from other `setup()`/`init()` functions.

Reply:
we could set up a AD1PCFG for our specific pins for UART however it isn't entirely necessary since AD1PCFG is specifically for the pins labeled as AN0 - AN12. Since the pins I am using 
(_TRISB6 = 0;  // Sets pin 15 (RB6) to U1TX output pin
    _TRISB10 = 1; // Sets pin 21 (RB10) to U1RX input pin
)
are not on any AN pins of the PIC24, it most likely isn't necessary to do any AD1PCFG stuff. At least for UART. At least thats what I understood from my TA. He did suggest though that if we aren't doing any analog readings, we could just set all the pins to digital just in case.

<img width="684" alt="Screenshot 2024-04-08 at 1 37 04 PM" src="https://github.com/cecilsmith/GPS-MINPOINT/assets/163371180/51ae0407-23ac-4080-beac-576fcef8d47b">